crates/polars-core/src/frame/hash_join/single_keys.rs

```diff
@@ -20,32 +20,26 @@ where
                     PlHashMap::with_capacity(HASHMAP_INIT_SIZE);
 
                 let n_partitions = n_partitions as u64;
-                let mut offset = 0;
-                for keys in &keys {
-                    let keys = keys.as_ref();
-                    let len = keys.len() as IdxSize;
 
-                    let mut cnt = 0;
-                    keys.iter().for_each(|k| {
-                        let idx = cnt + offset;
-                        cnt += 1;
-
-                        if this_partition(k.as_u64(), partition_no, n_partitions) {
-                            let entry = hash_tbl.entry(*k);
+                keys.iter()
+                    .flat_map(|array| array.as_ref().iter())
```

Comment:
Can you revert these changes. Flatmap is slower than the code we have as it has in indirection the double join doesn't have. 

I rather have you don't change code that is related to performance of joins in a PR that fixes the join validation.

Reply:
Reverted. 

However, I wonder why using `flat_map` is slower. Any suggestions?

Reply:
Every iteration you do a loop over the nested iterators: https://doc.rust-lang.org/src/core/iter/adapters/flatten.rs.html#476